### DISCUSSION TITLE
Completions don't work for test parameters

### DISCUSSION THREAD

**Discussion by rchiodo** [microsoft/pylance-release#4388] [Enhancement]
Example code:

```python
import pytest

@pytest.fixture
def test_fixture() -> List[int]:
    return [4]

def test_func(test_fixture):
      test_fixture. # No completions here
```

We could probably implement first level completions? If the type is any and its declaration is a parameter, get the type from the extensions?

**Comment by erictraut** [comment]
I think this is worth considering. Instead of `Any`, you should check for `Unknown`, since that's what pyright infers for params with no annotation.

**Comment by ciszko** [comment]
With the recent VSCode IntelliSense improvements for pytest, I've started searching if it's a bug (at least it looks like a bug for someone from the outside). `test_fixture` as an argument is properly hinted but then inside of the test it's resolved as `Any`. I just want to say that this suggested enhancement would be a lifesaver for anyone working extensively with fixtures.

**Comment by debonte** [comment]
@ciszko, please file a new issue on this and include some sample code to demonstrate the behavior that you want.

**Comment by ciszko** [comment]
Do I need to create another issue if this one already exists? It's exactly what  @rchiodo wrote.
A fixture is hinted as an argument
![image](https://user-images.githubusercontent.com/45177564/226303504-6dd9ed9f-3aca-4fc1-a5bd-5ee965017277.png)
But then if I were to use it, it's `Any`
![image](https://user-images.githubusercontent.com/45177564/226303753-b10a7450-d7d7-430a-a154-13b62889259e.png)

**Comment by night-as** [comment]
Any progress regarding this issue ? It would be really helpful.

**Comment by rchiodo** [comment]
@night-as this is waiting for more people to up vote it, so no it's not on our backlog yet.

**Comment by rchiodo** [comment]
Moving to discussions to hopefully get more visibility.

**Comment by matifriaz** [comment]
Any update on this issue?

**Comment by y2kbugger** [comment]
Not just completions, but I think it makes sense to have type checking as well, not sure it that all comes along together within pyright or not

**Comment by YakovL** [comment]
What are the workarounds you guys currently use for this? I have to admit, it's quite annoying. It's also useful to check [the discussion on the Pytest end](https://github.com/pytest-dev/pytest/issues/3834) (I tend to believe it's more of their issue than that of Pylance). I've found a fragile workaround:
```
@pytest.fixture(name='session_manager')
def get_session_manager():
    return GameSessionsManager()

def test_create_session(session_manager: GameSessionsManager):
    ...
```
The problem about it is casting the `GameSessionsManager` type to the argument: once implementation of `get_session_manager` is changed (say, returns `GameSessionsManager | None`), things can get broken.

**Reply by heejaechang** [reply]
latest pylance should now support completion for fixture without explicitly typing parameters.

![image](https://github.com/microsoft/pylance-release/assets/1333179/d8325fe5-b312-4dbb-bf0c-91d6b60cce3d)

the grey color type annotation is inlay hint not explicit typing.

**Reply by YakovL** [reply]
Wow, that's true even without type hint for the argument. However, without type hint it still complains that the param type is unknown. Would be great if that one is handled, too!

![image](https://github.com/microsoft/pylance-release/assets/1131924/a63be882-044a-40ec-bae1-41cc171e3dcb)

**Reply by heejaechang** [reply]
Those pytest IDE features, such as hover and completion, are best-effort basis and that's why we support them. However, diagnostics are a different story. We will still show those errors if you are using strict mode, and the only way to get rid of them will be by explicitly adding type hints, as you used to.

**Reply by RamiAwar** [reply]
I don't get autocomplete still actually. I'm using latest Pylance, I can see the correct type when I hover over the parameter. But upon using the parameter inside the function, it has type Unknown. It's weird..

**Comment by Olkesandr** [comment]
Any chance to be fixed?

**Reply by rchiodo** [reply]
This should be working already. It does for me with this example:

![image](https://github.com/user-attachments/assets/859900fd-909c-4f0f-b227-bd48964bbf3f)

You have to turn on pytest support:

![image](https://github.com/user-attachments/assets/566d7ec9-11ca-4100-b41d-8c9690e1412c)

**Reply by Olkesandr** [reply]
Thanks for your reply. I did some testing and found that it doesn't work when test function is inside a class:
![image](https://github.com/user-attachments/assets/26b18598-72e1-4ae5-94cb-d43808645285)

**Reply by rchiodo** [reply]
Thanks, I added an issue for that:
https://github.com/microsoft/pylance-release/issues/7351
